lib/hl7.ex

```diff
@@ -2,14 +2,201 @@ defmodule HL7 do
   @moduledoc """
   Utility functions to load HL7 files as local streams.
   """
-  require Logger
 
-  @buffer_size 32768
+  defstruct segments: []
 
+  @buffer_size 32768
   @type file_type_hl7 :: :mllp | :line | nil
 
+  @type hl7_map_data() :: %{
+          optional(non_neg_integer) => hl7_map_data() | String.t(),
+          e: non_neg_integer()
+        }
+  @type hl7_list_data() :: String.t() | [hl7_list_data]
+
+  @type segment() :: %{
+          0 => String.t(),
+          optional(pos_integer) => hl7_map_data() | String.t(),
+          e: non_neg_integer()
+        }
+
+  @type t() :: %__MODULE__{segments: [segment()]}
+
+  @type parsed_hl7_segments :: t() | [segment()]
+  @type parsed_hl7 :: t() | segment() | [segment()] | hl7_map_data()
+
+  alias HL7.Path
+
+  @doc ~S"""
+  Creates an HL7 struct from HL7 data (accepting text, lists or the deprecated `HL7.Message` struct).
+
+  The segment maps using integer keys corresponding to the data's HL7 positions (starting at 1).
+  Segment names are stored at position 0. To save on space and to make a cleaner presentation, empty values
+  are not included in the output. Instead, each map contains an `:e` field that notes the highest index
+  present in the source data.
+  """
+  @spec parse!(hl7_list_data() | String.t() | HL7.Message.t()) :: t()
+  def parse!(segments) when is_list(segments) do
+    segments = Enum.map(segments, fn segment -> to_map(%{}, 0, segment) end)
+    %__MODULE__{segments: segments}
+  end
+
+  def parse!(text) when is_binary(text) do
+    text |> HL7.Message.to_list() |> parse!()
+  end
+
+  def parse!(%HL7.Message{} = message) do
+    message |> HL7.Message.to_list() |> parse!()
+  end
+
+  @spec parse(String.t()) :: {:ok, t()} | {:error, HL7.InvalidMessage.t()}
+  def parse(text) when is_binary(text) do
+    case HL7.Message.new(text) do
```

Comment:
It is strange to me that HL7.Message is deprecated, but we are using it here in the new function.

Reply:
We'll have HL7.Message for a while, as might others, and this provides an easy transition between the two for now until we deprecate/rm it in the future.